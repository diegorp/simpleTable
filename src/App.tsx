import groceryItems from './data/groceryItems';
import Table from './components/Table';

function App() {
  return (
    <main className='flex flex-col h-screen w-screen p-20'>
      <Table
        title="Today's groceries"
        data={groceryItems}
        rowIdKey='id'
        pagination
        sort
        hideRowId
        filter
      />
    </main>
  );
}

export default App;
