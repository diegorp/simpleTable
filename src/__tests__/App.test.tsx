import { describe, it, expect, beforeEach, vi } from 'vitest';
import { render } from '@testing-library/react';
import App from '@/App';

beforeEach(() => {
  HTMLElement.prototype.hidePopover = vi.fn();
  HTMLElement.prototype.showPopover = vi.fn();
});

describe('<App />', () => {
  it('should match snapshot', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('should render the table component', () => {
    const { getByRole } = render(<App />);

    const table = getByRole('table');
    expect(table).toBeInTheDocument();
  });
});
