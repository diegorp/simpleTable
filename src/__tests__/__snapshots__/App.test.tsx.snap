// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<App /> > should match snapshot 1`] = `
<div>
  <main
    class="flex flex-col h-screen w-screen p-20"
  >
    <article
      class="flex flex-col gap-4 overflow-hidden"
    >
      <header
        class="flex items-center justify-between"
      >
        <h2
          class="text-[15px]"
        >
          Today's groceries
        </h2>
        <button
          aria-label="Filter"
          class="flex items-center gap-2 h-[32px] text-[13px] px-4 py-2 rounded cursor-pointer bg-transparent border border-[#E2E8F0] hover:bg-gray-100 transition-colors duration-200 ease-in-out "
          popovertarget="filter-popover"
          popovertargetaction="toggle"
          style="anchor-name: --filter-btn;"
        >
          <svg
            fill="#939BAA"
            height="12"
            viewBox="0 0 24 24"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1,5 C1,4.44771525 1.44266033,4 1.99895656,4 L3.00104344,4 C3.55275191,4 4,4.44386482 4,5 C4,5.55228475 3.55733967,6 3.00104344,6 L1.99895656,6 C1.44724809,6 1,5.55613518 1,5 Z M12,5 C12,4.44771525 12.444837,4 12.9955775,4 L22.0044225,4 C22.5542648,4 23,4.44386482 23,5 C23,5.55228475 22.555163,6 22.0044225,6 L12.9955775,6 C12.4457352,6 12,5.55613518 12,5 Z M8,6 C7.44771525,6 7,5.55228475 7,5 C7,4.44771525 7.44771525,4 8,4 C8.55228475,4 9,4.44771525 9,5 C9,5.55228475 8.55228475,6 8,6 Z M8,8 C6.34314575,8 5,6.65685425 5,5 C5,3.34314575 6.34314575,2 8,2 C9.65685425,2 11,3.34314575 11,5 C11,6.65685425 9.65685425,8 8,8 Z M1,19 C1,18.4477153 1.44266033,18 1.99895656,18 L3.00104344,18 C3.55275191,18 4,18.4438648 4,19 C4,19.5522847 3.55733967,20 3.00104344,20 L1.99895656,20 C1.44724809,20 1,19.5561352 1,19 Z M12,19 C12,18.4477153 12.444837,18 12.9955775,18 L22.0044225,18 C22.5542648,18 23,18.4438648 23,19 C23,19.5522847 22.555163,20 22.0044225,20 L12.9955775,20 C12.4457352,20 12,19.5561352 12,19 Z M8,20 C7.44771525,20 7,19.5522847 7,19 C7,18.4477153 7.44771525,18 8,18 C8.55228475,18 9,18.4477153 9,19 C9,19.5522847 8.55228475,20 8,20 Z M8,22 C6.34314575,22 5,20.6568542 5,19 C5,17.3431458 6.34314575,16 8,16 C9.65685425,16 11,17.3431458 11,19 C11,20.6568542 9.65685425,22 8,22 Z M1,12 C1,11.4477153 1.4556644,11 1.99539757,11 L10.0046024,11 C10.5543453,11 11,11.4438648 11,12 C11,12.5522847 10.5443356,13 10.0046024,13 L1.99539757,13 C1.44565467,13 1,12.5561352 1,12 Z M19,12 C19,11.4477153 19.4433532,11 20.0093689,11 L21.9906311,11 C22.5480902,11 23,11.4438648 23,12 C23,12.5522847 22.5566468,13 21.9906311,13 L20.0093689,13 C19.4519098,13 19,12.5561352 19,12 Z M15,13 C14.4477153,13 14,12.5522847 14,12 C14,11.4477153 14.4477153,11 15,11 C15.5522847,11 16,11.4477153 16,12 C16,12.5522847 15.5522847,13 15,13 Z M15,15 C13.3431458,15 12,13.6568542 12,12 C12,10.3431458 13.3431458,9 15,9 C16.6568542,9 18,10.3431458 18,12 C18,13.6568542 16.6568542,15 15,15 Z"
            />
          </svg>
           
          <span>
            Filter
          </span>
        </button>
        <div
          aria-label="Filter table data by category and value"
          id="filter-popover"
          popover=""
          role="dialog"
          style="position-anchor: --filter-btn; translate: -100%;"
        >
          <section
            class="flex flex-col gap-2 p-4 border border-[#E2E8F0] rounded shadow-lg bg-white z-30 mt-1 w-[250px]"
          >
            <div
              class="flex gap-2 shrink-0"
            >
              <span
                class="text-sm font-medium shrink-0"
              >
                Filter by:
              </span>
              <select
                aria-label="Select a category for filtering"
                class="shrink-0 cursor-pointer"
              >
                <option
                  selected=""
                  value=""
                >
                  Select a category
                </option>
                <option
                  value="name"
                >
                  name
                </option>
                <option
                  value="section"
                >
                  section
                </option>
                <option
                  value="price"
                >
                  price
                </option>
                <option
                  value="weight"
                >
                  weight
                </option>
              </select>
            </div>
            <input
              aria-label="Filter value..."
              placeholder="Filter value..."
              type="text"
              value=""
            />
            <button
              aria-label="Apply filter"
              class="px-3 py-1 border border-neutral-300 text-black rounded transition-colors duration-200 ease-in-out bg-transparent hover:bg-neutral-100 disabled:opacity-50 cursor-pointer disabled:hover:bg-transparent disabled:cursor-not-allowed"
              disabled=""
            >
              Apply
            </button>
            <button
              aria-label="Clear filter"
              class="px-3 py-1 bg-neutral-800 border border-neutral-800 text-white rounded transition-colors duration-200 ease-in-out hover:bg-neutral-900 cursor-pointer"
            >
              Clear
            </button>
          </section>
        </div>
      </header>
      <div
        class="flex-1 overflow-auto border border-[#E2E8F0]"
      >
        <table
          class="min-w-full table-fixed overflow-auto text-gray-800"
        >
          <thead>
            <tr>
              <th
                aria-sort="other"
                class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
                scope="col"
              >
                <button
                  aria-label="Sort by name"
                  class="w-full h-full px-6 py-3 flex items-center gap-2 capitalize cursor-pointer text-left"
                >
                  name
                </button>
              </th>
              <th
                aria-sort="other"
                class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
                scope="col"
              >
                <button
                  aria-label="Sort by section"
                  class="w-full h-full px-6 py-3 flex items-center gap-2 capitalize cursor-pointer text-left"
                >
                  section
                </button>
              </th>
              <th
                aria-sort="other"
                class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
                scope="col"
              >
                <button
                  aria-label="Sort by price"
                  class="w-full h-full px-6 py-3 flex items-center gap-2 capitalize cursor-pointer text-left"
                >
                  price
                </button>
              </th>
              <th
                aria-sort="other"
                class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
                scope="col"
              >
                <button
                  aria-label="Sort by weight"
                  class="w-full h-full px-6 py-3 flex items-center gap-2 capitalize cursor-pointer text-left"
                >
                  weight
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            class="bg-white"
          >
            <tr
              class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
            >
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Bananas
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Produce
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                1.29
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                0.5
              </td>
            </tr>
            <tr
              class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
            >
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Apples
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Produce
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                2.49
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                1
              </td>
            </tr>
            <tr
              class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
            >
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Milk
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Dairy
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                3.99
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                1
              </td>
            </tr>
            <tr
              class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
            >
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Eggs
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Dairy
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                2.19
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                0.5
              </td>
            </tr>
            <tr
              class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
            >
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Bread
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                Bakery
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                2.99
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
              >
                0.5
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <nav
        aria-label="Table pagination"
        class="flex shrink-0 gap-12 w-full text-sm text-gray-500 items-center justify-start"
      >
        <div
          class="flex items-center gap-2"
        >
          <span>
            Rows per page: 
          </span>
          <select
            aria-label="Rows per page"
            aria-valuenow="5"
          >
            <option
              value="5"
            >
              5
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
            <option
              value="100"
            >
              100
            </option>
          </select>
        </div>
        <div
          class="flex items-center gap-2"
        >
          <span
            role="status"
          >
            1
             -
             
            5
             of
             
            56
          </span>
          <button
            aria-label="Previous page"
            class="text-xl cursor-pointer mr-2 disabled:opacity-50 disabled:cursor-not-allowed"
            disabled=""
          >
            ‹
          </button>
          <button
            aria-label="Next page"
            class="text-xl cursor-pointer disabled:opacity-50 disabled:cursor-not-allowed"
          >
            ›
          </button>
        </div>
      </nav>
    </article>
  </main>
</div>
`;
